dm/syncer/data_validator.go

```diff
@@ -581,55 +598,67 @@ func (v *DataValidator) startValidateWorkers() {
 	}
 
 	for _, tblChange := range v.loadedPendingChanges {
-		for key, row := range tblChange.rows {
+		for key, row := range tblChange.jobs {
 			v.dispatchRowChange(key, row)
 		}
 	}
 }
 
-func (v *DataValidator) dispatchRowChange(key string, row *rowChange) {
+func (v *DataValidator) dispatchRowChange(key string, row *rowChangeJob) {
 	hashVal := int(utils.GenHashKey(key)) % v.workerCnt
 	v.workers[hashVal].rowChangeCh <- row
 }
 
-func (v *DataValidator) genValidateTableInfo(sourceTable *filter.Table) (*validateTableInfo, error) {
+func (v *DataValidator) genValidateTableInfo(sourceTable *filter.Table, columnCount int) (*validateTableInfo, error) {
 	targetTable := v.syncer.route(sourceTable)
 	// there are 2 cases tracker may drop table:
 	// 1. checkpoint rollback, tracker may recreate tables and drop non-needed tables
 	// 2. when operate-schema
 	// in case 1, we add another layer synchronization to make sure we don't get a dropped table when recreation.
 	// 	for non-needed tables, we will not validate them.
 	// in case 2, validator should be paused
+	res := &validateTableInfo{targetTable: targetTable}
 	tableInfo, err := v.syncer.getTrackedTableInfo(sourceTable)
 	if err != nil {
-		return nil, err
+		if schema.IsTableNotExists(err) {
+			// not a table need to sync
+			res.message = tableNotSyncedOrDropped
+			return res, nil
+		}
+		return res, err
 	}
-	table := &validateTableInfo{
-		Source: sourceTable,
-		Info:   tableInfo,
-		Target: targetTable,
+	if len(tableInfo.Columns) < columnCount {
+		res.message = moreColumnInBinlogMsg
+		return res, nil
 	}
 
-	var primaryIdx *model.IndexInfo
-	for _, idx := range tableInfo.Indices {
-		if idx.Primary {
-			primaryIdx = idx
+	tableID := utils.GenTableID(targetTable)
+	downstreamTableInfo, err := v.syncer.getDownStreamTableInfo(v.tctx, tableID, tableInfo)
+	if err != nil {
+		// todo: might be connection error, then return error, or downstream table not exists, then set state to stopped.
+		return res, err
+	}
+	pk := downstreamTableInfo.WhereHandle.UniqueNotNullIdx
+	if pk == nil {
+		res.message = tableWithoutPrimaryKeyMsg
+		return res, nil
+	}
+	// offset of pk column is adjusted using source table info, the offsets should stay in range of ev.ColumnCount.
+	for _, col := range pk.Columns {
+		if col.Offset >= columnCount {
+			res.message = downstreamPKColumnOutOfBoundsMsg
+			return res, nil
 		}
 	}
-	// todo: PKIsHandle = true when table has primary key like "id int primary key CLUSTERED", since schema-tracker(we get from it)
-	// todo: only use downstream DDL when the task is incremental only, will support this case later.
-	if primaryIdx == nil {
-		// return partial initialized result
-		return table, errNoPrimaryKey
+	// if current TI has more columns, clone and strip columns
+	if len(tableInfo.Columns) > columnCount {
+		tableInfo = tableInfo.Clone()
+		tableInfo.Columns = tableInfo.Columns[:columnCount]
```

Comment:
Considering the following case:
upstream binlog has columns: `a`, `b`, `c`; downstream table has column: `b`, `c`, `d`, `e` (this may happen because we don't forbid DDL event currently)

Simply slicing the `Columns` doesn't work. Should we leave a warning when DDLs come and tries modifying the schemas? Or should we have more decent way to check if the schemas matched?

Reply:
downstream should have all the columns as in upstream, else syncer will fail too.

we need all columns in the binlog not dropped or switch order in tracker ddl, it's a constraint.

Reply:
syncer have a hidden feature which may cause different column name between upstream and down stream  see more `s.mappingDML` 😵

Reply:
it's not changing column name but values(add prefix or suffix). would not support tasks with this hidden feature in this version.